File: src/intake_rules.py
import time
from datetime import date, datetime, timezone


AUTHORIZED_APPROVERS = {"Transition Manager", "Security Officer"}
APPROVE_ACTION = "Aprovar Intake"


def _utc_now(now: datetime | None = None) -> datetime:
    if now is None:
        return datetime.now(timezone.utc)
    if now.tzinfo is None:
        return now.replace(tzinfo=timezone.utc)
    return now.astimezone(timezone.utc)


def _timestamp_utc(value: datetime) -> str:
    return value.isoformat().replace("+00:00", "Z")


def persist_audit_event(audit_store: list, audit_event: dict) -> dict:
    started_at = time.perf_counter()
    audit_store.append(audit_event)
    elapsed_seconds = time.perf_counter() - started_at

    return {
        "persisted": True,
        "elapsed_seconds": elapsed_seconds,
    }


def validate_dr_evidence_date(value: str, now: datetime | None = None) -> dict:
    reference_date = _utc_now(now).date()

    try:
        evidence_date = date.fromisoformat(value)
    except ValueError:
        return {
            "is_valid": False,
            "message": "Data da evidência de Disaster Recovery inválida.",
            "age_days": None,
        }

    age_days = (reference_date - evidence_date).days

    if age_days < 0:
        return {
            "is_valid": False,
            "message": "Data da evidência de Disaster Recovery não pode ser futura.",
            "age_days": age_days,
        }

    if age_days > 365:
        return {
            "is_valid": False,
            "message": "Evidência de Disaster Recovery fora do prazo aceite.",
            "age_days": age_days,
        }

    return {
        "is_valid": True,
        "message": "Evidência de Disaster Recovery válida.",
        "age_days": age_days,
    }


def approve_intake(
    role: str,
    user_id: str,
    current_status: str,
    is_intake_valid: bool,
    now: datetime | None = None,
) -> dict:
    timestamp = _utc_now(now)

    if role not in AUTHORIZED_APPROVERS:
        return {
            "allowed": False,
            "status": current_status,
            "message": "Acesso Negado",
            "audit_events": [
                {
                    "user_id": user_id,
                    "action": APPROVE_ACTION,
                    "timestamp_utc": _timestamp_utc(timestamp),
                }
            ],
        }

    if not is_intake_valid:
        return {
            "allowed": False,
            "status": current_status,
            "message": "Intake inválido.",
            "audit_events": [],
        }

    return {
        "allowed": True,
        "status": "Aprovado",
        "message": "Intake aprovado.",
        "audit_events": [],
    }

File: LovableApp/scope-prototype-main/src/lib/intakeRules.ts
type DrEvidenceDateResult = {
  valid: boolean;
  error: string | null;
};

type ApprovalAttemptInput = {
  role: string;
  userId: string;
  currentStatus: string;
  now: Date;
};

type AuditEvent = {
  userId: string;
  action: "APPROVE_INTAKE";
  timestampUtc: string;
};

type ApprovalAttemptResult = {
  approved: boolean;
  nextStatus: string;
  feedback: string;
  auditEvent: AuditEvent | null;
};

const DAY_IN_MS = 24 * 60 * 60 * 1000;
const APPROVAL_AUTHORIZED_ROLES = ["Transition Manager", "Security Officer"];

const toUtcDateOnly = (date: Date) =>
  Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate());

export const validateDrEvidenceDate = (
  evidenceDate: string,
  now: Date,
): DrEvidenceDateResult => {
  if (!evidenceDate) {
    return { valid: false, error: "Data do Teste de DR é obrigatória." };
  }

  const parsedDate = new Date(evidenceDate);

  if (Number.isNaN(parsedDate.getTime())) {
    return { valid: false, error: "Data inválida." };
  }

  const evidenceDay = toUtcDateOnly(parsedDate);
  const currentDay = toUtcDateOnly(now);

  if (evidenceDay > currentDay) {
    return { valid: false, error: "A data não pode ser futura." };
  }

  const ageInDays = (currentDay - evidenceDay) / DAY_IN_MS;

  if (ageInDays > 365) {
    return { valid: false, error: "Evidência fora do prazo (>365 dias)." };
  }

  return { valid: true, error: null };
};

export const isOverrideJustificationValid = (justification: string): boolean =>
  justification.length >= 20;

export const evaluateApprovalAttempt = ({
  role,
  userId,
  currentStatus,
  now,
}: ApprovalAttemptInput): ApprovalAttemptResult => {
  if (!APPROVAL_AUTHORIZED_ROLES.includes(role)) {
    return {
      approved: false,
      nextStatus: currentStatus,
      feedback:
        "Acesso Negado: a sua role não tem permissão para aprovar este Intake.",
      auditEvent: {
        userId,
        action: "APPROVE_INTAKE",
        timestampUtc: now.toISOString(),
      },
    };
  }

  return {
    approved: true,
    nextStatus: "Aprovado",
    feedback: "Intake aprovado com sucesso.",
    auditEvent: null,
  };
};
